src/Pact/Eval.hs

```diff
@@ -458,22 +565,30 @@ evaluateDefs info mdef defs = do
       evalError i $ "Recursion detected: " <> prettyList pl
 
   -- the order of evaluation matters for 'dresolve' - this *must* be a left fold
-  let dresolve ds (d,dn,_) = HM.insert dn (Ref $ unify ds <$> d) ds
-      unifiedDefs = foldl' dresolve HM.empty sortedDefs
-
-  traverse (runSysOnly . evalConsts) unifiedDefs
+  -- Todo: Charge gas here.
+  isExecutionFlagSet FlagDisableInlineMemCheck >>= \case
+    True -> do
+      let dresolve ds (d,dn,_) = HM.insert dn (Ref $ unify ds <$> d) ds
+          unifiedDefs = (foldl' dresolve HM.empty sortedDefs)
+      traverse (runSysOnly . evalConsts) unifiedDefs
+    False -> do
+      hl <- fromIntegral <$> view (eeGasEnv.geGasLimit)
+      hf <- foldlM (dresolveMem hl info) (HeapFold HM.empty M.empty 0 0) sortedDefs
+      -- note: hfLastGas computes post const reduction so we wet the gas to the amount currently used
+      -- it would error if it was past the gas limit
+      evalGas .= (_hfLastGas hf)
+      pure (_hfAllDefs hf)
   where
     mkSomeDoc = either (SomeDoc . pretty) (SomeDoc . pretty)
-
     -- | traverse to find deps and form graph
-    traverseGraph ds memo = fmap stronglyConnCompR $ forM (HM.toList ds) $ \(dn,d) -> do
-      d' <- forM d $ \(f :: Name) -> do
+    traverseGraph allDefs memo = fmap stronglyConnCompR $ forM (HM.toList allDefs) $ \(defName,defTerm) -> do
```

Comment:
renamed these only because `ds`, `dn` and `d` were a bit hard to track what was what when working with this.

Alternatively a type sig would've helped but it's sort of big.